build/utils/npm.go

```diff
@@ -9,189 +9,284 @@ import (
+	if err != nil {
+		return nil, err
 	}
-
-	return TraverseDependencies(dependenciesMap, traverseDependenciesFunc, threads)
-}
-
-func TraverseDependencies(dependenciesMap map[string]*entities.Dependency, traverseDependenciesFunc func(dependency *entities.Dependency) (bool, error), threads int) (dependenciesList []entities.Dependency, err error) {
-	producerConsumer := parallel.NewBounedRunner(threads, false)
-	dependenciesChan := make(chan *entities.Dependency)
-	errorChan := make(chan error, 1)
-
-	go func() {
-		defer producerConsumer.Done()
-		for _, dep := range dependenciesMap {
-			handlerFunc := createHandlerFunc(dep, dependenciesChan, traverseDependenciesFunc)
-			_, err = producerConsumer.AddTaskWithError(handlerFunc, func(err error) {
-				// Write the error to the channel, but don't wait if the channel buffer is full.
-				select {
-				case errorChan <- err:
-				default:
-					return
-				}
-			})
-			if err != nil {
-				return
-			}
+	var missingPeerDeps, missingBundledDeps []string
+	for _, dep := range prototypeTree {
+		if dep.npmLsDependency.Integrity == "" && dep.npmLsDependency.InBundle {
+			missingBundledDeps = append(missingBundledDeps, dep.Id)
+			continue
 		}
-	}()
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-	go func() {
-		for dep := range dependenciesChan {
-			dependenciesList = append(dependenciesList, *dep)
+		if dep.npmLsDependency.Integrity == "" && len(dep.PeerMissing) > 0 {
+			missingPeerDeps = append(missingPeerDeps, dep.Id)
+			continue
+		}
+		dep.Md5, dep.Sha1, dep.Sha256, err = calculateChecksum(cacache, dep.Name, dep.Version, dep.Integrity, log)
+		if err != nil {
+			// Here, we don't know where is the tarball (or if it is actually exists in the filesystem) so we can't calculate the dependency checksum.
+			// This case happends when the package-lock.json with property '"lockfileVersion": 1,' gets updated to version '"lockfileVersion": 2,' (from npm v6 to npm v7/v8).
+			// Seems like the compatibility upgrades may result in dependencies losing their integrity.
+			// We use the integrity to get's the dependencies tarball
+			log.Error("couldn't calculate checksum for : '" + dep.Id + "'. Try to delete 'node_models' and/or 'package-lock.json'.")
+			return nil, err
 		}
-		wg.Done()
-	}()
 
-	producerConsumer.Run()
-	if err != nil {
-		return
+		dependenciesList = append(dependenciesList, dep.Dependency)
 	}
-	close(dependenciesChan)
-	wg.Wait()
-
-	// Read the error from the channel, but don't wait if the channel buffer is empty.
-	select {
-	case err = <-errorChan:
-	default:
-		return
+	if len(missingPeerDeps) > 0 {
+		log.Info("The following dependencies will not be included in the build-info, because the 'npm ls' command did not return its integrity.\nThe reason why the version wasn't returned may be because the package is a 'peerDependency', which was not manually installed.\n It is therefore okay to skip this dependency: \n" + strings.Join(missingPeerDeps, "\n"))
```

Comment:
```suggestion
		log.Info("The following dependencies will not be included in the build-info, because the 'npm ls' command did not return their integrity.\nThe reason why the version wasn't returned may be because the package is a 'peerDependency', which was not manually installed.\n It is therefore okay to skip this dependency: \n" + strings.Join(missingPeerDeps, "\n"))
```